subsystem/cluster_test.go

```diff
@@ -176,41 +177,74 @@ var _ = Describe("Cluster tests", func() {
+	}
+
+	return false, swag.StringValue(c.Status)
+}
+
 func waitForClusterState(ctx context.Context, clusterID strfmt.UUID, state string, timeout time.Duration, stateInfo string) {
 	log.Infof("Waiting for cluster %s status %s", clusterID, state)
-	for start := time.Now(); time.Since(start) < timeout; {
-		rep, err := userBMClient.Installer.GetCluster(ctx, &installer.GetClusterParams{ClusterID: clusterID})
-		Expect(err).NotTo(HaveOccurred())
-		c := rep.GetPayload()
-		if swag.StringValue(c.Status) == state {
-			break
+	var lastState string = ""
+	var success bool
+
+	for start, successInRow := time.Now(), 0; time.Since(start) < timeout; {
+		success, lastState = isClusterInState(ctx, clusterID, state, stateInfo)
+
+		if success {
+			successInRow++
+		} else {
+			successInRow = 0
 		}
+
+		// Wait for cluster state to be consistent
+		if successInRow >= minSuccessesInRow {
+			return
+		}
+
 		time.Sleep(time.Second)
 	}
-	rep, err := userBMClient.Installer.GetCluster(ctx, &installer.GetClusterParams{ClusterID: clusterID})
+
+	Fail(fmt.Sprintf("Cluster %s wasn't in state %s for %d times in a row. Last state was %s",
+		clusterID, state, minSuccessesInRow, lastState))
+}
+
+func isHostInState(ctx context.Context, clusterID strfmt.UUID, hostID strfmt.UUID, state string) (bool, string) {
+	rep, err := userBMClient.Installer.GetHost(ctx, &installer.GetHostParams{ClusterID: clusterID, HostID: hostID})
 	Expect(err).NotTo(HaveOccurred())
 	c := rep.GetPayload()
-	Expect(swag.StringValue(c.Status)).Should(Equal(state))
-	if stateInfo != IgnoreStateInfo {
-		Expect(swag.StringValue(c.StatusInfo)).Should(Equal(stateInfo))
-	}
+	return swag.StringValue(c.Status) == state, swag.StringValue(c.Status)
 }
 
 func waitForHostState(ctx context.Context, clusterID strfmt.UUID, hostID strfmt.UUID, state string, timeout time.Duration) {
 	log.Infof("Waiting for host %s state %s", hostID, state)
-	for start := time.Now(); time.Since(start) < timeout; {
-		rep, err := userBMClient.Installer.GetHost(ctx, &installer.GetHostParams{ClusterID: clusterID, HostID: hostID})
-		Expect(err).NotTo(HaveOccurred())
-		c := rep.GetPayload()
-		if swag.StringValue(c.Status) == state {
-			break
+	var lastState string = ""
+	var success bool
+
+	for start, successInRow := time.Now(), 0; time.Since(start) < timeout; {
+		success, lastState = isHostInState(ctx, clusterID, hostID, state)
+
+		if success {
+			successInRow++
+		} else {
+			successInRow = 0
 		}
+
+		// Wait for host state to be consistent
+		if successInRow >= minSuccessesInRow {
+			return
+		}
+
 		time.Sleep(time.Second)
 	}
-	rep, err := userBMClient.Installer.GetHost(ctx, &installer.GetHostParams{ClusterID: clusterID, HostID: hostID})
-	Expect(err).NotTo(HaveOccurred())
-	c := rep.GetPayload()
-	ExpectWithOffset(1, swag.StringValue(c.Status)).Should(Equal(state))
+
+	Fail(fmt.Sprintf("Host %s in Cluster %s wasn't in state %s for %d times in a row. Last state was %s",
```

Comment:
Please change it to Expect the diff will be clearer and no one will need to read the message.